config/hcl2shim/paths.go

```diff
@@ -48,6 +51,86 @@ func RequiresReplace(attrs []string, ty cty.Type) ([]cty.Path, error) {
 	return paths, nil
 }
 
+func AttributePathToLegacyPath(path string) (string, []error) {
+	if path == "" || strings.HasSuffix(path, ".#") || strings.HasSuffix(path, ".%") {
+		return path, []error{}
+	}
+
+	ret := []string{}
+
+	expr, diags := hcl2syntax.ParseExpression([]byte(path), "", hcl2.Pos{Line: 1, Column: 1})
```

Comment:
It looks like your goal here is to extract the sequence of traversal steps, in which case a more direct way to do that would be:

```go
    traversal, diags := hcl2syntax.ParseTraversalAbs([]byte(path), "", hcl2.Pos{Line: 1, Column: 1})
    if diags.HasErrors() {
        return "", diags.Errs()
    }
    for _, step := range traversal {
        // .. then type-switch step to handle TraverseRoot, TraverseAttr, and TraverseIndex
        // for formatting purposes.
    }
```

With that said, I find this function a little confusing: it seems like it's trying to go from a "new-style" (HCL2) path syntax to the old-style `foo.bar.0.baz`-type path, but yet above it accepts paths ending in `.#` and `.%` which are legacy-style flatmap paths already. For shims in here I think it's best for them to be very clear about what they are mapping from and to, since code using them gets very confusing already and so them having descriptive, unambiguous names helps to understand how values are flowing from one representation to another and what we have at each point in the program.

Reply:
Thanks for all the early feedback.

> With that said, I find this function a little confusing

You're right. It is indeed confusing because this function is called from _many_ places with old syntax (from `schema.ResourceData`, effectively from all providers), not just from `schema.schemaMap` validation logic, which is much easier to refactor alone.

One better/cleaner approach I can think of would be to change behaviour of all relevant `ResourceConfig` methods (`CheckSet`, `Get`, `GetRaw`, `IsComputed`, `IsSet`) to just accept `cty.Path` and expose new methods `LegacyCheckSet`, `LegacyGet`, `LegacyGetRaw`, `LegacyIsComputed`, `LegacyIsSet` with existing string-based argument which can be used throughout the rest of the codebase. It is technically a breaking change, but (hopefully) only on package-level.

The other (cleanest?) approach would be to refactor everything between `schema.Provider.Validate` and `schemaMap.get` to `Validate(v cty.Value)` and get rid of all `ResourceConfig` references there, which is the approach I will most likely take. 😃